configuration.md
---
# !!! DO NOT MODIFY !!!
# This file is auto-generated by the gendoc tool based on the source code of the app.
description: This section describes the configuration parameters and their types for INX-Tendercoo.
keywords:
- IOTA Node 
- Hornet Node
- Coordinator
- Configuration
- JSON
- Customize
- Config
- reference
---


# Core Configuration

INX-Tendercoo uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `inx-tendercoo` executable.

For example:
```bash
inx-tendercoo -c config_defaults.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
inx-tendercoo -h --full
```

## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "5m"          |
| [log](#app_shutdown_log) | Configuration for log                                                                                  | object |               |

### <a id="app_shutdown_log"></a> Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "5m",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name                                     | Description                                                                 | Type    | Default value |
| ---------------------------------------- | --------------------------------------------------------------------------- | ------- | ------------- |
| level                                    | The minimum enabled logging level                                           | string  | "info"        |
| disableCaller                            | Stops annotating logs with the calling function's file name and line number | boolean | true          |
| disableStacktrace                        | Disables automatic stacktrace capturing                                     | boolean | false         |
| stacktraceLevel                          | The level stacktraces are captured and above                                | string  | "panic"       |
| encoding                                 | The logger's encoding (options: "json", "console")                          | string  | "console"     |
| [encodingConfig](#logger_encodingconfig) | Configuration for encodingConfig                                            | object  |               |
| outputPaths                              | A list of URLs, file paths or stdout/stderr to write logging output to      | array   | stdout        |
| disableEvents                            | Prevents log messages from being raced as events                            | boolean | true          |

### <a id="logger_encodingconfig"></a> EncodingConfig

| Name        | Description                                                                                                | Type   | Default value |
| ----------- | ---------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| timeEncoder | Sets the logger's timestamp encoding. (options: "nanos", "millis", "iso8601", "rfc3339" and "rfc3339nano") | string | "rfc3339"     |

Example:

```json
  {
    "logger": {
      "level": "info",
      "disableCaller": true,
      "disableStacktrace": false,
      "stacktraceLevel": "panic",
      "encoding": "console",
      "encodingConfig": {
        "timeEncoder": "rfc3339"
      },
      "outputPaths": [
        "stdout"
      ],
      "disableEvents": true
    }
  }
```

## <a id="inx"></a> 3. INX

| Name                  | Description                                                                                        | Type   | Default value    |
| --------------------- | -------------------------------------------------------------------------------------------------- | ------ | ---------------- |
| address               | The INX address to which to connect to                                                             | string | "localhost:9029" |
| maxConnectionAttempts | The amount of times the connection to INX will be attempted before it fails (1 attempt per second) | uint   | 30               |
| targetNetworkName     | The network name on which the node should operate on (optional)                                    | string | ""               |

Example:

```json
  {
    "inx": {
      "address": "localhost:9029",
      "maxConnectionAttempts": 30,
      "targetNetworkName": ""
    }
  }
```

## <a id="coordinator"></a> 4. Coordinator

| Name                                  | Description                                                     | Type    | Default value |
| ------------------------------------- | --------------------------------------------------------------- | ------- | ------------- |
| interval                              | Target interval in which milestones are issued                  | string  | "5s"          |
| maxTrackedBlocks                      | Maximum number of blocks tracked by the milestone tip selection | int     | 10000         |
| whiteFlagParentsSolidTimeout          | Timeout for the ComputeWhiteFlag INX call                       | string  | "2s"          |
| singleNode                            | Set to true for single node test networks                       | boolean | false         |
| [tipsel](#coordinator_tipsel)         | Configuration for tipsel                                        | object  |               |
| [tendermint](#coordinator_tendermint) | Configuration for tendermint                                    | object  |               |

### <a id="coordinator_tipsel"></a> Tipsel

| Name                     | Description                                                                                              | Type   | Default value |
| ------------------------ | -------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| maxTips                  | Maximum number of tips returned                                                                          | int    | 7             |
| reducedConfirmationLimit | Stop selection, when tips reference less additional blocks than this fraction (compared to the best tip) | float  | 0.6           |
| timeout                  | Timeout after which tip selection is canceled                                                            | string | "100ms"       |

### <a id="coordinator_tendermint"></a> Tendermint

| Name                                             | Description                                                                 | Type   | Default value     |
| ------------------------------------------------ | --------------------------------------------------------------------------- | ------ | ----------------- |
| bindAddress                                      | Bind address for incoming connections                                       | string | "0.0.0.0:26656"   |
| consensusPrivateKey                              | Node's private key used for consensus                                       | string | ""                |
| nodePrivateKey                                   | Node's private key used for P2P communication                               | string | ""                |
| root                                             | Root folder to store config and database                                    | string | "tendermint"      |
| logLevel                                         | Logging level of the Tendermint Core; cannot be lower than the global level | string | "info"            |
| genesisTime                                      | Time the blockchain started or will start in Unix time using seconds        | int    | 0                 |
| chainID                                          | Identifier of the blockchain; every chain must have a unique ID             | string | "tendercoo"       |
| maxRetainBlocks                                  | Maximum number of blocks to keep in the database (0=no pruning)             | uint   | 0                 |
| peers                                            | Addresses of the Tendermint nodes to connect to (ID@Host:Port)              | array  |                   |
| [consensus](#coordinator_tendermint_consensus)   | Configuration for consensus                                                 | object |                   |
| [prometheus](#coordinator_tendermint_prometheus) | Configuration for prometheus                                                | object |                   |
| validators                                       | Set of validators                                                           | object | see example below |

### <a id="coordinator_tendermint_consensus"></a> Consensus

| Name                      | Description                                                            | Type    | Default value |
| ------------------------- | ---------------------------------------------------------------------- | ------- | ------------- |
| createEmptyBlocks         | Whether empty blocks are created                                       | boolean | false         |
| createEmptyBlocksInterval | Create empty blocks after waiting this long without receiving anything | string  | "0s"          |
| blockInterval             | Delay between blocks                                                   | string  | "700ms"       |
| skipBlockTimeout          | Make progress as soon as we have all the precommits                    | boolean | false         |

### <a id="coordinator_tendermint_prometheus"></a> Prometheus

| Name        | Description                              | Type    | Default value     |
| ----------- | ---------------------------------------- | ------- | ----------------- |
| enabled     | Toggle for Prometheus metrics collection | boolean | false             |
| bindAddress | Prometheus listening address binding     | string  | "localhost:26660" |

Example:

```json
  {
    "coordinator": {
      "interval": "5s",
      "maxTrackedBlocks": 10000,
      "whiteFlagParentsSolidTimeout": "2s",
      "singleNode": false,
      "tipsel": {
        "maxTips": 7,
        "reducedConfirmationLimit": 0.6,
        "timeout": "100ms"
      },
      "tendermint": {
        "bindAddress": "0.0.0.0:26656",
        "consensusPrivateKey": "",
        "nodePrivateKey": "",
        "root": "tendermint",
        "logLevel": "info",
        "genesisTime": 0,
        "chainID": "tendercoo",
        "maxRetainBlocks": 0,
        "peers": null,
        "consensus": {
          "createEmptyBlocks": false,
          "createEmptyBlocksInterval": "0s",
          "blockInterval": "700ms",
          "skipBlockTimeout": false
        },
        "prometheus": {
          "enabled": false,
          "bindAddress": "localhost:26660"
        },
        "validators": null
      }
    }
  }
```

## <a id="profiling"></a> 5. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling component is enabled        | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```



documentation/docs/configuration.md
---
# !!! DO NOT MODIFY !!!
# This file is auto-generated by the gendoc tool based on the source code of the app.
description: This section describes the configuration parameters and their types for INX-Tendercoo.
keywords:
- IOTA Node 
- Hornet Node
- Coordinator
- Configuration
- JSON
- Customize
- Config
- reference
---


# Core Configuration

INX-Tendercoo uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `inx-tendercoo` executable.

For example:
```bash
inx-tendercoo -c config_defaults.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
inx-tendercoo -h --full
```

## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "5m"          |
| [log](#app_shutdown_log) | Configuration for log                                                                                  | object |               |

### <a id="app_shutdown_log"></a> Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "5m",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name              | Description                                                                 | Type    | Default value |
| ----------------- | --------------------------------------------------------------------------- | ------- | ------------- |
| level             | The minimum enabled logging level                                           | string  | "info"        |
| disableCaller     | Stops annotating logs with the calling function's file name and line number | boolean | true          |
| disableStacktrace | Disables automatic stacktrace capturing                                     | boolean | false         |
| stacktraceLevel   | The level stacktraces are captured and above                                | string  | "panic"       |
| encoding          | The logger's encoding (options: "json", "console")                          | string  | "console"     |
| outputPaths       | A list of URLs, file paths or stdout/stderr to write logging output to      | array   | stdout        |
| disableEvents     | Prevents log messages from being raced as events                            | boolean | true          |

Example:

```json
  {
    "logger": {
      "level": "info",
      "disableCaller": true,
      "disableStacktrace": false,
      "stacktraceLevel": "panic",
      "encoding": "console",
      "outputPaths": [
        "stdout"
      ],
      "disableEvents": true
    }
  }
```

## <a id="inx"></a> 3. INX

| Name                  | Description                                                                                        | Type   | Default value    |
| --------------------- | -------------------------------------------------------------------------------------------------- | ------ | ---------------- |
| address               | The INX address to which to connect to                                                             | string | "localhost:9029" |
| maxConnectionAttempts | The amount of times the connection to INX will be attempted before it fails (1 attempt per second) | uint   | 30               |
| targetNetworkName     | The network name on which the node should operate on (optional)                                    | string | ""               |

Example:

```json
  {
    "inx": {
      "address": "localhost:9029",
      "maxConnectionAttempts": 30,
      "targetNetworkName": ""
    }
  }
```

## <a id="coordinator"></a> 4. Coordinator

| Name                                  | Description                                                     | Type    | Default value |
| ------------------------------------- | --------------------------------------------------------------- | ------- | ------------- |
| interval                              | Target interval in which milestones are issued                  | string  | "5s"          |
| maxTrackedBlocks                      | Maximum number of blocks tracked by the milestone tip selection | int     | 10000         |
| whiteFlagParentsSolidTimeout          | Timeout for the ComputeWhiteFlag INX call                       | string  | "2s"          |
| singleNode                            | Set to true for single node test networks                       | boolean | false         |
| [tipsel](#coordinator_tipsel)         | Configuration for tipsel                                        | object  |               |
| [tendermint](#coordinator_tendermint) | Configuration for tendermint                                    | object  |               |

### <a id="coordinator_tipsel"></a> Tipsel

| Name                     | Description                                                                                              | Type   | Default value |
| ------------------------ | -------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| maxTips                  | Maximum number of tips returned                                                                          | int    | 7             |
| reducedConfirmationLimit | Stop selection, when tips reference less additional blocks than this fraction (compared to the best tip) | float  | 0.6           |
| timeout                  | Timeout after which tip selection is canceled                                                            | string | "100ms"       |

### <a id="coordinator_tendermint"></a> Tendermint

| Name                                             | Description                                                                 | Type   | Default value     |
| ------------------------------------------------ | --------------------------------------------------------------------------- | ------ | ----------------- |
| bindAddress                                      | Bind address for incoming connections                                       | string | "0.0.0.0:26656"   |
| consensusPrivateKey                              | Node's private key used for consensus                                       | string | ""                |
| nodePrivateKey                                   | Node's private key used for P2P communication                               | string | ""                |
| root                                             | Root folder to store config and database                                    | string | "tendermint"      |
| logLevel                                         | Logging level of the Tendermint Core; cannot be lower than the global level | string | "info"            |
| genesisTime                                      | Time the blockchain started or will start in Unix time using seconds        | int    | 0                 |
| chainID                                          | Identifier of the blockchain; every chain must have a unique ID             | string | "tendercoo"       |
| maxRetainBlocks                                  | Maximum number of blocks to keep in the database (0=no pruning)             | uint   | 0                 |
| peers                                            | Addresses of the Tendermint nodes to connect to (ID@Host:Port)              | array  |                   |
| [consensus](#coordinator_tendermint_consensus)   | Configuration for consensus                                                 | object |                   |
| [prometheus](#coordinator_tendermint_prometheus) | Configuration for prometheus                                                | object |                   |
| validators                                       | Set of validators                                                           | object | see example below |

### <a id="coordinator_tendermint_consensus"></a> Consensus

| Name                      | Description                                                            | Type    | Default value |
| ------------------------- | ---------------------------------------------------------------------- | ------- | ------------- |
| createEmptyBlocks         | Whether empty blocks are created                                       | boolean | false         |
| createEmptyBlocksInterval | Create empty blocks after waiting this long without receiving anything | string  | "0s"          |
| blockInterval             | Delay between blocks                                                   | string  | "700ms"       |
| skipBlockTimeout          | Make progress as soon as we have all the precommits                    | boolean | false         |

### <a id="coordinator_tendermint_prometheus"></a> Prometheus

| Name        | Description                              | Type    | Default value     |
| ----------- | ---------------------------------------- | ------- | ----------------- |
| enabled     | Toggle for Prometheus metrics collection | boolean | false             |
| bindAddress | Prometheus listening address binding     | string  | "localhost:26660" |

Example:

```json
  {
    "coordinator": {
      "interval": "5s",
      "maxTrackedBlocks": 10000,
      "whiteFlagParentsSolidTimeout": "2s",
      "singleNode": false,
      "tipsel": {
        "maxTips": 7,
        "reducedConfirmationLimit": 0.6,
        "timeout": "100ms"
      },
      "tendermint": {
        "bindAddress": "0.0.0.0:26656",
        "consensusPrivateKey": "",
        "nodePrivateKey": "",
        "root": "tendermint",
        "logLevel": "info",
        "genesisTime": 0,
        "chainID": "tendercoo",
        "maxRetainBlocks": 0,
        "peers": null,
        "consensus": {
          "createEmptyBlocks": false,
          "createEmptyBlocksInterval": "0s",
          "blockInterval": "700ms",
          "skipBlockTimeout": false
        },
        "prometheus": {
          "enabled": false,
          "bindAddress": "localhost:26660"
        },
        "validators": null
      }
    }
  }
```

## <a id="profiling"></a> 5. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling plugin is enabled           | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```



components/decoo/params.go
package decoo

import (
	"time"

	"github.com/iotaledger/hive.go/app"
)

// ParametersDefinition contains the definition of configuration parameters used by the plugin.
type ParametersDefinition struct {
	Interval                     time.Duration `default:"5s" usage:"target interval in which milestones are issued"`
	MaxTrackedBlocks             int           `default:"10000" usage:"maximum number of blocks tracked by the milestone tip selection"`
	WhiteFlagParentsSolidTimeout time.Duration `default:"2s" usage:"timeout for the ComputeWhiteFlag INX call"`
	SingleNode                   bool          `default:"false" usage:"set to true for single node test networks"`

	TipSel struct {
		MaxTips                  int           `default:"7" usage:"maximum number of tips returned"`
		ReducedConfirmationLimit float64       `default:"0.6" usage:"stop selection, when tips reference less additional blocks than this fraction (compared to the best tip)"`
		Timeout                  time.Duration `default:"100ms" usage:"timeout after which tip selection is canceled"`
	} `name:"tipsel"`

	Tendermint struct {
		BindAddress         string `default:"0.0.0.0:26656" usage:"bind address for incoming connections"`
		ConsensusPrivateKey string `usage:"node's private key used for consensus"`
		NodePrivateKey      string `usage:"node's private key used for P2P communication"`

		Root            string `default:"tendermint" usage:"root folder to store config and database"`
		LogLevel        string `default:"info" usage:"logging level of the Tendermint Core; cannot be lower than the global level"`
		GenesisTime     int64  `default:"0" usage:"time the blockchain started or will start in Unix time using seconds"`
		ChainID         string `default:"tendercoo" usage:"identifier of the blockchain; every chain must have a unique ID"`
		MaxRetainBlocks uint   `default:"0" usage:"maximum number of blocks to keep in the database (0=no pruning)"`

		Peers      []string                    `usage:"addresses of the Tendermint nodes to connect to (ID@Host:Port)"`
		Validators map[string]ValidatorsConfig `noflag:"true" usage:"set of validators"`

		Consensus struct {
			CreateEmptyBlocks         bool          `default:"false" usage:"whether empty blocks are created"`
			CreateEmptyBlocksInterval time.Duration `default:"0s" usage:"create empty blocks after waiting this long without receiving anything"`
			BlockInterval             time.Duration `default:"700ms" usage:"delay between blocks"`
			SkipBlockTimeout          bool          `default:"false" usage:"make progress as soon as we have all the precommits"`
		}

		Prometheus struct {
			Enabled     bool   `default:"false" usage:"toggle for Prometheus metrics collection"`
			BindAddress string `default:"localhost:26660" usage:"Prometheus listening address binding"`
		}
	}
}

// ValidatorsConfig defines the config options for one validator.
type ValidatorsConfig struct {
	PublicKey string `usage:"consensus key of the validator"`
	Power     int64  `usage:"voting power of the validator"`
}

// Parameters contains the configuration parameters of the decoo app.
var Parameters = &ParametersDefinition{}

var params = &app.ComponentParams{
	Params: map[string]any{
		"coordinator": Parameters,
	},
}
